homeassistant/components/azure_event_hub/config_flow.py

```diff
@@ -0,0 +1,208 @@
+    CONF_EVENT_HUB_NAMESPACE,
+    CONF_EVENT_HUB_SAS_KEY,
+    CONF_EVENT_HUB_SAS_POLICY,
+    CONF_MAX_DELAY,
+    CONF_SEND_INTERVAL,
+    CONF_USE_CONN_STRING,
+    DEFAULT_OPTIONS,
+    DOMAIN,
+    STEP_CONN_STRING,
+    STEP_SAS,
+    STEP_USER,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+BASE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_EVENT_HUB_INSTANCE_NAME): str,
+        vol.Optional(CONF_USE_CONN_STRING, default=False): bool,
+    }
+)
+
+CONN_STRING_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_EVENT_HUB_CON_STRING): str,
+    }
+)
+
+SAS_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_EVENT_HUB_NAMESPACE): str,
+        vol.Required(CONF_EVENT_HUB_SAS_POLICY): str,
+        vol.Required(CONF_EVENT_HUB_SAS_KEY): str,
+    }
+)
+
+
+async def validate_data(data: dict[str, Any]) -> dict[str, str] | None:
+    """Validate the input."""
+    client = AzureEventHubClient.from_input(**data)
+    try:
+        await client.test_connection()
+    except EventHubError:
+        return {"base": "cannot_connect"}
+    except Exception as exc:  # pylint: disable=broad-except
+        _LOGGER.warning("Unknown error: %s", exc)
+        return {"base": "unknown"}
+    return None
+
+
+class AEHConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for azure event hub."""
+
+    VERSION: int = 1
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self._data: dict[str, Any] = {}
+        self._options: dict[str, Any] = DEFAULT_OPTIONS
+        self._conn_string: bool | None = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return AEHOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input: dict[str, Any] = None) -> FlowResult:
+        """Handle the initial user step."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+        if user_input is None:
+            return self.async_show_form(step_id=STEP_USER, data_schema=BASE_SCHEMA)
+
+        self._conn_string = user_input.pop(CONF_USE_CONN_STRING)
+        self._data = user_input
+
+        if self._conn_string:
+            return await self.async_step_conn_string()
+        return await self.async_step_sas()
+
+    async def async_step_conn_string(
+        self, user_input: dict[str, Any] = None
+    ) -> FlowResult:
+        """Handle the connection string steps."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id=STEP_CONN_STRING,
+                data_schema=CONN_STRING_SCHEMA,
+                description_placeholders=self._data[CONF_EVENT_HUB_INSTANCE_NAME],
+                last_step=True,
+            )
+        errors = await self.async_update_and_validate_data(user_input)
+        if errors is not None:
+            return self.async_show_form(
```

Comment:
I'd put this form at the end of the function and nest the create entry call instead. Then we don't need two form calls in the step.